pkg/lib/synthesize/resource_getters.go

```diff
@@ -0,0 +1,301 @@
+	apiregistrationv1beta1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1beta1"
+
+	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+)
+
+const (
+	CSVKind                = "ClusterServiceVersion"
+	CRDKind                = "CustomResourceDefinition"
+	SecretKind             = "Secret"
+	ConfigMapKind          = "ConfigMap"
+	ClusterRoleKind        = "ClusterRole"
+	ClusterRoleBindingKind = "ClusterRoleBinding"
+	ServiceAccountKind     = "ServiceAccount"
+	ServiceKind            = "Service"
+	RoleKind               = "Role"
+	RoleBindingKind        = "RoleBinding"
+	PrometheusRuleKind     = "PrometheusRule"
+	ServiceMonitorKind     = "ServiceMonitor"
+	DeploymentKind         = "Deployment"
+	APIServiceKind         = "APIService"
+)
+
+// registerGVK registers the GVK of the kube resource we use in the CSV, explicitly including all versions.
+func registerGVK(scheme *runtime.Scheme) {
+	scheme.AddKnownTypes(operatorsv1alpha1.SchemeGroupVersion, &operatorsv1alpha1.ClusterServiceVersion{})
+
+	scheme.AddKnownTypes(extv1.SchemeGroupVersion, &extv1.CustomResourceDefinition{})
+	scheme.AddKnownTypes(extv1beta1.SchemeGroupVersion, &extv1beta1.CustomResourceDefinition{})
+
+	scheme.AddKnownTypes(cv1.SchemeGroupVersion,
+		&cv1.ServiceAccount{},
+		&cv1.ConfigMap{},
+		&cv1.Secret{})
+
+	scheme.AddKnownTypes(rv1.SchemeGroupVersion,
+		&rv1.Role{},
+		&rv1.RoleBinding{},
+		&rv1.ClusterRole{},
+		&rv1.ClusterRoleBinding{})
+
+	scheme.AddKnownTypes(appsv1.SchemeGroupVersion, &appsv1.Deployment{})
+
+	scheme.AddKnownTypes(apiregistrationv1.SchemeGroupVersion, &apiregistrationv1.APIService{})
+	scheme.AddKnownTypes(apiregistrationv1beta1.SchemeGroupVersion, &apiregistrationv1beta1.APIService{})
+}
+
+type GroupVersionKindName struct {
+	Group   string
+	Version string
+	Kind    string
+	Name    string
+}
+
+func NewGVKN(gvk schema.GroupVersionKind, name string) GroupVersionKindName {
+	return GroupVersionKindName{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind,
+		Name:    name,
+	}
+}
+
+func (gvkn *GroupVersionKindName) gvk() GroupVersionKind {
+	return GroupVersionKind{
+		Group:   gvkn.Group,
+		Version: gvkn.Version,
+		Kind:    gvkn.Kind,
+	}
+}
+
+type GroupVersionKind schema.GroupVersionKind
+
+type customResourceDefinition struct {
+	v1      extv1.CustomResourceDefinition
+	v1beta1 extv1beta1.CustomResourceDefinition
+	gvkn    GroupVersionKindName
+}
+
+func (crd *customResourceDefinition) getResourceGVKNs() (gvkns []GroupVersionKindName) {
+	switch crd.gvkn.Version {
+	case crd.v1.GroupVersionKind().Version:
+		for _, version := range crd.v1.Spec.Versions {
+			gvkns = append(gvkns, GroupVersionKindName{
+				Group:   crd.v1.Spec.Group,
+				Kind:    crd.v1.Spec.Names.Kind,
+				Version: version.Name,
+				Name:    crd.v1.GetName(),
+			})
+		}
+	case crd.v1beta1.GroupVersionKind().Version:
+		for _, version := range crd.v1beta1.Spec.Versions {
```

Comment:
note: for `v1beta1` CRDs there is also a `Version` field at the top-level which defines a singular version of the CRD. The `Versions` field is optional for `v1beta1` CRDs if a singular `Version` is defined if you check out the comments in the spec https://github.com/exdx/apiextensions-apiserver/blob/a30f5f256c16bc2a087771a4369ba63b8ed96b2a/pkg/apis/apiextensions/v1beta1/types.go#L79

So I think we may also need to check for that case here

Reply:
I did see that in another implementation, should have thought about it here. thanks for catching it